File: commands/Command31.kt
package com.mbadin.mmibasic.commands

import com.mbadin.mmibasic.CommandModel
import com.mbadin.mmibasic.toBinary

// 31 38 C7 07 01 00 C9 0C 1B 0C 03 00 00 00 0D 00 00 00 31 00 00 00 C1 00 00 00 07 03 00 00 18 1C 00 00 60 01 02 18 0D 37 B0 62 39 08 F7 36 01 02 14 17 23 04 D7 39 08 F7 37 01 02
// 31 0D F2 60 01 00 72 2B 07 05 00 42 7F 40 00 89
// 31 0D F2 61 01 00 78 2B 07 05 3E 0D 02 FD 08 FA
// 31 0D F2 6C 01 00 64 2B 07 05 01 01 79 05 03 91
// 31 0D F2 6D 01 00 6A 2B 07 05 3E 41 41 41 3E A0
// 31 09 F6 6E 01 00 70 2E 04 01 09 F3
// 31 0D F2 6F 01 00 72 2B 07 05 00 42 7F 40 00 86
// 31 0D F2 70 01 00 78 2B 07 05 00 42 7F 40 0D 02
// 31 0D F2 7A 01 00 64 2B 07 05 01 01 79 05 03 87
// 31 0D F2 7B 01 00 6A 2B 07 05 3E 41 41 41 3E B6
// 31 0F F0 DB 01 00 01 02 07 ED 01 00 C9 29 1B 0C 00 00
// 31 0A F5 02 01 00 77 00 07 02 00 00 BF
// 31 0C F3 04 01 00 7A 02 07 04 7F 09 09 01 CE
// 31 1E E1 09 01 01 5D 46 09 0B FF 01 FF 01 FB 01 83 01 FB 01 FF 01 83 01 EB 01 F7 01 FF 01 FF 01 3C
// 31 0D F2 0B 01 00 69 47 07 05 00 42 7F 40 00 95
// 31 0D F2 0C 01 00 6F 47 07 05 06 49 49 49 3E 98
// 31 09 F6 0D 01 00 75 4A 04 01 09 F1
// 31 1E E1 33 01 01 5D 46 09 0B FF 01 FF 01 FB 01 83 01 FB 01 FF 01 83 01 EB 01 F7 01 FF 01 FF 01 06
// 31 0D F2 35 01 00 69 47 07 05 00 42 7F 40 00 AB
// 31 0D F2 36 01 00 6F 47 07 05 06 49 49 49 3E A2
// 31 09 F6 37 01 00 75 4A 04 0D 02 FD

@ExperimentalUnsignedTypes
class Command31(data: UByteArray) : CommandModel(0x31.toUByte(), data) {
    /**
     * 0x00 - both colors
     * 0x01 - red color if bit is 1
     * 0x02 - red color if bit is 0
     */
    fun getColorType(): UByte {
        return getData()[2]
    }

    // [x, y]
    fun getCoords(): Array<UByte> {
        return arrayOf(getData()[3], getData()[4])
    }

    // [height, width]
    fun getSize(): Array<UByte> {
        return arrayOf(getData()[5], getData()[6])
    }

    fun getBytes(): List<String> {
        val width = getSize()[1]

        val bytes = getData().copyOfRange(7, getData().count() - 1)

        val count = bytes.count() / width.toInt()

        return Array(width.toInt()) { it }.map { index ->
            val start = count * index
            val end = start + count

            bytes
                .copyOfRange(start, end)
                .joinToString("") { it.toBinary().reversed() }
        }
    }
}

File: example/src/Command31.ts
import { ASCII } from "./ASCII";
import { Command } from "./Command";

// 0x31 0x0F 0xF0 0x0D 0x01 0x00 0xAE 0x02 0x07 0x07 0x7F 0x20 0x10 0x08 0x10 0x20 0x7F 0x66

type Options = {
  scale?: number;
};

export class Command31 extends Command {
  static readonly ID = "0x31";
  private payload: Uint8Array;

  static readonly COLOR_BOTH = 0x00;
  static readonly COLOR_RED = 0x01;
  static readonly COLOR_BLACK = 0x02;

  static readonly COLORS = {
    RED: "#ff0000",
    BLACK: "#000000",
  } as const;

  constructor(frame: string[]) {
    super(frame);

    this.payload = this.getPayload();
  }

  private getColorType() {
    return this.payload[2];
  }

  private getDimensions() {
    return {
      x: this.payload[3],
      y: this.payload[4],
      height: this.payload[5],
      width: this.payload[6],
    };
  }

  // Convert the pixel bits in the payload into an ASCII table string.
  toBitmap(bytes: Uint8Array, height: number, width: number) {
    const count = Math.floor(bytes.length / width) || 0;

    const maxRows = Math.min(height, count * 8);
    const rows = Array.from({ length: maxRows }, () => "");

    for (let col = 0; col < width; col++) {
      const start = col * count;
      let colBits = "";
      for (let b = 0; b < count; b++) {
        const byte = bytes[start + b] ?? 0;
        let bin = byte.toString(2).padStart(8, "0");
        bin = bin.split("").reverse().join("");
        colBits += bin;
      }

      rows[col] = colBits;
    }

    return rows
      .reduce<string[][]>((acc, row) => {
        const rowBits = row.split("");

        rowBits.forEach((bit, index) => {
          acc[index] = acc[index] ?? [];
          acc[index]?.push(bit);
        });

        return acc;
      }, [])
      .map((rowBits) => rowBits.join(""));
  }

  render(canvas: HTMLCanvasElement, opts?: Options) {
    this.isValid();

    const colorType = this.getColorType();
    const { x, y, height, width } = this.getDimensions();

    if (
      typeof colorType === "undefined" ||
      typeof x === "undefined" ||
      typeof y === "undefined" ||
      typeof height === "undefined" ||
      typeof width === "undefined"
    ) {
      throw new Error("Invalid Command 31 data");
    }

    const bytes = this.payload.slice(7, this.payload.length); // exclude final PLC-like byte

    const char = ASCII.bitmapToChar(this.toBitmap(bytes, height, width));

    if (char === "█") {
      console.warn(
        "No matching ASCII character found for bitmap, rendering as block:",
        {
          bitmap: this.toBitmap(bytes, height, width),
          raw: this.toRawArray(bytes),
        },
      );
    }

    // number of byte-rows per column
    const count = Math.floor(bytes.length / width) || 0;
    const scale = opts?.scale ?? 3;
    const ctx = canvas.getContext("2d");

    // helper to set one pixel (scaled rectangle)
    function setPixel(px: number, py: number, color: string) {
      if (ctx) {
        ctx.fillStyle = color;
        ctx.fillRect(px * scale, py * scale, scale, scale);
      }
    }

    // build columns
    for (let col = 0; col < width; col++) {
      const start = col * count;
      let colBits = "";
      for (let b = 0; b < count; b++) {
        const byte = bytes[start + b] ?? 0;
        let bin = byte.toString(2).padStart(8, "0");
        bin = bin.split("").reverse().join(""); // reversed as in Kotlin
        colBits += bin;
      }

      // render rows up to declared height
      for (let row = 0; row < Math.min(height, colBits.length); row++) {
        const bit = colBits[row];
        const xPos = x + col;
        const yPos = y + row;

        if (colorType === Command31.COLOR_BOTH) {
          // both colors: '1' -> red, '0' -> black
          if (bit === "1") setPixel(xPos, yPos, Command31.COLORS.RED);
          else setPixel(xPos, yPos, Command31.COLORS.BLACK);
        } else if (colorType === Command31.COLOR_RED) {
          if (bit === "1") setPixel(xPos, yPos, Command31.COLORS.RED);
        } else if (colorType === Command31.COLOR_BLACK) {
          if (bit === "1") setPixel(xPos, yPos, Command31.COLORS.BLACK);
        }
      }
    }
  }
}
